### PR TITLE
Adds Julia as a plugin target for generating parsers

### PR DESCRIPTION
**Pull request by davidmclifton**
Tokenizer, Parser and some working example files. Additional details shared via email to Dmitry.

The generated Julia parser has one dependency - the very common DataStructures package which provides Queue and Stack implementations.

### CODE REVIEW

File: src/plugins/julia/lr/lr-parser-generator-julia.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+ */
+const JL_LR_PARSER_TEMPLATE = fs.readFileSync(
+  `${__dirname}/../templates/lr.template.jl`,
+  'utf-8',
+);
+
+/**
+ * LR parser generator for Example language.
+ */
+export default class LRParserGeneratorExample extends LRParserGeneratorDefault {
```
Comment (DmitrySoshnikov) [inline src/plugins/julia/lr/lr-parser-generator-julia.js:32]: `LRParserGeneratorJulia`

File: src/plugins/julia/julia-parser-generator-trait.js [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+   * Generates parsing table in Julia language Map format.
+   *
+   * An array of records, where index - 1 is a state number, and a value is a
+   * dictionary from an encoded symbol (number) to a parsing action.
+   *
+   * The parsing action can be "Shift/s", "Reduce/r", a state
+   * transition number, or "Accept/acc".
+   *
+   * Example:
+   *
```
Comment (DmitrySoshnikov) [inline src/plugins/julia/julia-parser-generator-trait.js:39]: nit: remove empty example

File: src/plugins/julia/templates/lr.template.jl [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+struct SyntaxError <: Exception
+    msg::AbstractString
+end
+
+function Base.showerror(io::IO, err::SyntaxError)
+    print(io, err.msg)
+end
+
+Base.@kwdef mutable struct yyLoc
+    startoffset
```
Comment (DmitrySoshnikov) [inline src/plugins/julia/templates/lr.template.jl:47]: Does Julia have recommended way of naming fields/variables with multiple words? E.g. underscore_case for `start_offset` or camelCase for `startOffset`? Ditto for other variables/fields in these files.

File: src/plugins/julia/templates/lr.template.jl [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+# Tokenizer.
+
+{{{TOKENIZER}}}
+
+#=
+  Throws default "Unexpected token" exception, showing the actual
+  line from the source, pointing with the ^ marker to the bad token.
+  In addition, shows line:column location.
+=#
+function throwunexpectedtoken(tokenizerData::TokenizerData, symbol, line::Integer, column::Integer)
```
Comment (DmitrySoshnikov) [inline src/plugins/julia/templates/lr.template.jl:75]: Ditto for method names naming – is there a recommended format?

Also, see if this method should be in the tokenizer template/class.

File: src/plugins/julia/templates/lr.template.jl [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+    throw(SyntaxError(string("Incorrect Syntax\n\n", split(tokenizerData.initstring, "\n")[line], "\n", " "^(column - 1), "^\nUnexpected Token: \"", symbol, "\" at ", line, ":", column, ".")))
```
Comment (DmitrySoshnikov) [inline src/plugins/julia/templates/lr.template.jl:76]: We can keep the `"Incorrect Syntax"` (and in general custom error messages per plugin) if you'd like for Julia's output, however we should probably standardize the output to be the consistent with other Syntax plugins.

**Review by DmitrySoshnikov — COMMENTED** [review]
Great start! Some comments are inline:

File: src/plugins/julia/templates/lr.template.jl (hunk 3 above)
Comment (davidmclifton) [reply src/plugins/julia/templates/lr.template.jl:47]: This is a Julia oddity I came across - here's the style guide, refer to the section on "Use naming conventions consistent with Julia Base" which includes the note "modules and type names use capitalization and camel case: module SparseArrays, struct UnitRange" followed by "functions are lowercase and, when readable, with multiple words squashed together" - https://docs.julialang.org/en/v1/manual/style-guide/#Use-naming-conventions-consistent-with-Julia-base/ - this squash together bit seems to be consistently used for values as well within a type, such as variables defined in a struct or function.

File: src/plugins/julia/templates/lr.template.jl (hunk 4 above)
Comment (davidmclifton) [reply src/plugins/julia/templates/lr.template.jl:75]: See comment on startoffset/startOffset

Comment (davidmclifton) [reply src/plugins/julia/templates/lr.template.jl:75]: Will take a look at moving it over to tokenizer.

File: src/plugins/julia/lr/lr-parser-generator-julia.js (hunk 1 above)
Comment (davidmclifton) [reply src/plugins/julia/lr/lr-parser-generator-julia.js:32]: will update with fix

File: src/plugins/julia/julia-parser-generator-trait.js (hunk 2 above)
Comment (davidmclifton) [reply src/plugins/julia/julia-parser-generator-trait.js:39]: will remove

File: src/plugins/julia/templates/lr.template.jl (hunk 5 above)
Comment (davidmclifton) [reply src/plugins/julia/templates/lr.template.jl:76]: Removed the "Incorrect Syntax", I think this is aligned now with how the cs plugin reports errors.

**Comment by davidmclifton** [comment]
Committed changes per request comments

File: src/plugins/julia/templates/lr.template.jl [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+# parse methods in your use case, use 'import' instead of 'using'
+export parse
+
+# --------------------------------------------------------------
+# Shared includes
+using DataStructures
+
+# Constants and globals
+const EOF = "\$"
+yytext = ""
```
Comment (DmitrySoshnikov) [inline src/plugins/julia/templates/lr.template.jl:31]: Per discussion, see if we can remove this globals, and make them per parser/tokenizer data structure (as it's done in [other plugins](https://github.com/DmitrySoshnikov/syntax/blob/master/src/plugins/rust/rust-parser-generator-trait.js#L349)).  We can do this as an optimization in the follow up PRs.

**Review by DmitrySoshnikov — APPROVED** [review]
Looks great – thanks for the contribution!

**Comment by DmitrySoshnikov** [comment]
We will also need to update [main README](https://github.com/DmitrySoshnikov/syntax#language-agnostic-parser-generator) file to add Julia language, and also add a specific README to Julia plugin directory, similar to the [README from the Rust plugin](https://github.com/DmitrySoshnikov/syntax/blob/master/src/plugins/rust/README.md).

File: src/plugins/julia/templates/lr.template.jl (hunk 6 above)
Comment (davidmclifton) [reply src/plugins/julia/templates/lr.template.jl:31]: I will be working on this in the near future

**Pull request merged**